In [3]:
import pandas as pd
import numpy as np


from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report


from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import LabelEncoder

In [4]:
# Loading training dataset

In [5]:
train_data=pd.read_csv('C:/Users/NAMRATA/Downloads/DATA SCIENCE/ASSIGNMENT/Support Vector Machine/SalaryData_Train(1).csv')
train_data

age          workclass    education  educationno        maritalstatus  \
0       39          State-gov    Bachelors           13        Never-married   
1       50   Self-emp-not-inc    Bachelors           13   Married-civ-spouse   
2       38            Private      HS-grad            9             Divorced   
3       53            Private         11th            7   Married-civ-spouse   
4       28            Private    Bachelors           13   Married-civ-spouse   
...    ...                ...          ...          ...                  ...   
30156   27            Private   Assoc-acdm           12   Married-civ-spouse   
30157   40            Private      HS-grad            9   Married-civ-spouse   
30158   58            Private      HS-grad            9              Widowed   
30159   22            Private      HS-grad            9        Never-married   
30160   52       Self-emp-inc      HS-grad            9   Married-civ-spouse   

               occupation    relationship    race      sex  capitalgain  \
0            Adm-clerical   Not-in-family   White     Male         2174   
1         Exec-managerial         Husband   White     Male            0   
2       Handlers-cleaners   Not-in-family   White     Male            0   
3       Handlers-cleaners         Husband   Black     Male            0   
4          Prof-specialty            Wife   Black   Female            0   
...                   ...             ...     ...      ...          ...   
30156        Tech-support            Wife   White   Female            0   
30157   Machine-op-inspct         Husband   White     Male            0   
30158        Adm-clerical       Unmarried   White   Female            0   
30159        Adm-clerical       Own-child   White     Male            0   
30160     Exec-managerial            Wife   White   Female        15024   

       capitalloss  hoursperweek          native  Salary  
0                0            40   United-States   <=50K  
1                0            13   United-States   <=50K  
2                0            40   United-States   <=50K  
3                0            40   United-States   <=50K  
4                0            40            Cuba   <=50K  
...            ...           ...             ...     ...  
30156            0            38   United-States   <=50K  
30157            0            40   United-States    >50K  
30158            0            40   United-States   <=50K  
30159            0            20   United-States   <=50K  
30160            0            40   United-States    >50K  

[30161 rows x 14 columns]

In [6]:
train_data.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [7]:
labelencoder = LabelEncoder()
train_data.iloc[:, -1] = labelencoder.fit_transform(train_data.iloc[:,-1])
train_data

age          workclass    education  educationno        maritalstatus  \
0       39          State-gov    Bachelors           13        Never-married   
1       50   Self-emp-not-inc    Bachelors           13   Married-civ-spouse   
2       38            Private      HS-grad            9             Divorced   
3       53            Private         11th            7   Married-civ-spouse   
4       28            Private    Bachelors           13   Married-civ-spouse   
...    ...                ...          ...          ...                  ...   
30156   27            Private   Assoc-acdm           12   Married-civ-spouse   
30157   40            Private      HS-grad            9   Married-civ-spouse   
30158   58            Private      HS-grad            9              Widowed   
30159   22            Private      HS-grad            9        Never-married   
30160   52       Self-emp-inc      HS-grad            9   Married-civ-spouse   

               occupation    relationship    race      sex  capitalgain  \
0            Adm-clerical   Not-in-family   White     Male         2174   
1         Exec-managerial         Husband   White     Male            0   
2       Handlers-cleaners   Not-in-family   White     Male            0   
3       Handlers-cleaners         Husband   Black     Male            0   
4          Prof-specialty            Wife   Black   Female            0   
...                   ...             ...     ...      ...          ...   
30156        Tech-support            Wife   White   Female            0   
30157   Machine-op-inspct         Husband   White     Male            0   
30158        Adm-clerical       Unmarried   White   Female            0   
30159        Adm-clerical       Own-child   White     Male            0   
30160     Exec-managerial            Wife   White   Female        15024   

       capitalloss  hoursperweek          native  Salary  
0                0            40   United-States       0  
1                0            13   United-States       0  
2                0            40   United-States       0  
3                0            40   United-States       0  
4                0            40            Cuba       0  
...            ...           ...             ...     ...  
30156            0            38   United-States       0  
30157            0            40   United-States       1  
30158            0            40   United-States       0  
30159            0            20   United-States       0  
30160            0            40   United-States       1  

[30161 rows x 14 columns]

In [8]:
train_data_x=pd.get_dummies(train_data.iloc[:,:13])
train_data_x

age  educationno  capitalgain  capitalloss  hoursperweek  \
0       39           13         2174            0            40   
1       50           13            0            0            13   
2       38            9            0            0            40   
3       53            7            0            0            40   
4       28           13            0            0            40   
...    ...          ...          ...          ...           ...   
30156   27           12            0            0            38   
30157   40            9            0            0            40   
30158   58            9            0            0            40   
30159   22            9            0            0            20   
30160   52            9        15024            0            40   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
0                           0                     0                   0   
1                           0                     0                   0   
2                           0                     0                   1   
3                           0                     0                   1   
4                           0                     0                   1   
...                       ...                   ...                 ...   
30156                       0                     0                   1   
30157                       0                     0                   1   
30158                       0                     0                   1   
30159                       0                     0                   1   
30160                       0                     0                   0   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  ...  \
0                            0                            0  ...   
1                            0                            1  ...   
2                            0                            0  ...   
3                            0                            0  ...   
4                            0                            0  ...   
...                        ...                          ...  ...   
30156                        0                            0  ...   
30157                        0                            0  ...   
30158                        0                            0  ...   
30159                        0                            0  ...   
30160                        1                            0  ...   

       native_ Portugal  native_ Puerto-Rico  native_ Scotland  native_ South  \
0                     0                    0                 0              0   
1                     0                    0                 0              0   
2                     0                    0                 0              0   
3                     0                    0                 0              0   
4                     0                    0                 0              0   
...                 ...                  ...               ...            ...   
30156                 0                    0                 0              0   
30157                 0                    0                 0              0   
30158                 0                    0                 0              0   
30159                 0                    0                 0              0   
30160                 0                    0                 0              0   

       native_ Taiwan  native_ Thailand  native_ Trinadad&Tobago  \
0                   0                 0                        0   
1                   0                 0                        0   
2                   0                 0                        0   
3                   0                 0                        0   
4                   0                 0                        0   
...               ...               ...                      ...   
30156               0              

In [9]:
x=train_data_x
y=train_data['Salary']

In [10]:
# Splitting data into training and testing data set

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.3)

In [12]:
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((21112, 102), (21112,), (9049, 102), (9049,))

Tuning svm model with different values of parameters 'gmma' , 'c' and kernal='rbf' for best accuracy¶

In [13]:
?GridSearchCV

In [14]:
model= SVC()
param_grid = [{'kernel':['rbf'],'gamma':[50,5,10,0.5],'C':[0.1] }]
gsv = GridSearchCV(model,param_grid)
gsv.fit(x_train,y_train)

GridSearchCV(estimator=SVC(),
             param_grid=[{'C': [0.1], 'gamma': [50, 5, 10, 0.5],
                          'kernel': ['rbf']}])

In [15]:
# Accoriding to above gridsearch method these are the best parameters

In [16]:
gsv.best_params_ , gsv.best_score_

({'C': 0.1, 'gamma': 0.5, 'kernel': 'rbf'}, 0.7514209872107506)

In [17]:
# Building model with above parameteres

In [18]:
model_2=SVC(C=0.01,gamma=0.5,kernel='rbf')
model_2.fit(x_train , y_train)
y_pred = model_2.predict(x_test)
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)

Accuracy = 75.15747596419494


We got accurcy of 75.5% for the model with C=0.01,gamma=0.5 and kernal as rbf

Building model with c=0.01 and gamma as default

In [20]:
model_final = SVC(C=0.01,kernel='rbf')
model_final.fit(x_train , y_train)

SVC(C=0.01)

In [21]:
y_pred = model_final.predict(x_test)

In [22]:
confusion_matrix(y_test, y_pred)

array([[6779,   22],
       [1830,  418]], dtype=int64)

In [23]:
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)

Accuracy = 79.53365012708586


With above parameters we got better accuracy of 79.6% than compared with previous model.

Applying Prediction using above model for given test data

In [24]:
data_test=pd.read_csv('C:/Users/NAMRATA/Downloads/DATA SCIENCE/ASSIGNMENT/Support Vector Machine/SalaryData_Test(1).csv')
data_test

age      workclass      education  educationno        maritalstatus  \
0       25        Private           11th            7        Never-married   
1       38        Private        HS-grad            9   Married-civ-spouse   
2       28      Local-gov     Assoc-acdm           12   Married-civ-spouse   
3       44        Private   Some-college           10   Married-civ-spouse   
4       34        Private           10th            6        Never-married   
...    ...            ...            ...          ...                  ...   
15055   33        Private      Bachelors           13        Never-married   
15056   39        Private      Bachelors           13             Divorced   
15057   38        Private      Bachelors           13   Married-civ-spouse   
15058   44        Private      Bachelors           13             Divorced   
15059   35   Self-emp-inc      Bachelors           13   Married-civ-spouse   

               occupation    relationship                 race      sex  \
0       Machine-op-inspct       Own-child                Black     Male   
1         Farming-fishing         Husband                White     Male   
2         Protective-serv         Husband                White     Male   
3       Machine-op-inspct         Husband                Black     Male   
4           Other-service   Not-in-family                White     Male   
...                   ...             ...                  ...      ...   
15055      Prof-specialty       Own-child                White     Male   
15056      Prof-specialty   Not-in-family                White   Female   
15057      Prof-specialty         Husband                White     Male   
15058        Adm-clerical       Own-child   Asian-Pac-Islander     Male   
15059     Exec-managerial         Husband                White     Male   

       capitalgain  capitalloss  hoursperweek          native  Salary  
0                0            0            40   United-States   <=50K  
1                0            0            50   United-States   <=50K  
2                0            0            40   United-States    >50K  
3             7688            0            40   United-States    >50K  
4                0            0            30   United-States   <=50K  
...            ...          ...           ...             ...     ...  
15055            0            0            40   United-States   <=50K  
15056            0            0            36   United-States   <=50K  
15057            0            0            50   United-States   <=50K  
15058         5455            0            40   United-States   <=50K  
15059            0            0            60   United-States    >50K  

[15060 rows x 14 columns]

In [25]:
labelencoder = LabelEncoder()
data_test.iloc[:, -1] = labelencoder.fit_transform(data_test.iloc[:,-1])
data_test

age      workclass      education  educationno        maritalstatus  \
0       25        Private           11th            7        Never-married   
1       38        Private        HS-grad            9   Married-civ-spouse   
2       28      Local-gov     Assoc-acdm           12   Married-civ-spouse   
3       44        Private   Some-college           10   Married-civ-spouse   
4       34        Private           10th            6        Never-married   
...    ...            ...            ...          ...                  ...   
15055   33        Private      Bachelors           13        Never-married   
15056   39        Private      Bachelors           13             Divorced   
15057   38        Private      Bachelors           13   Married-civ-spouse   
15058   44        Private      Bachelors           13             Divorced   
15059   35   Self-emp-inc      Bachelors           13   Married-civ-spouse   

               occupation    relationship                 race      sex  \
0       Machine-op-inspct       Own-child                Black     Male   
1         Farming-fishing         Husband                White     Male   
2         Protective-serv         Husband                White     Male   
3       Machine-op-inspct         Husband                Black     Male   
4           Other-service   Not-in-family                White     Male   
...                   ...             ...                  ...      ...   
15055      Prof-specialty       Own-child                White     Male   
15056      Prof-specialty   Not-in-family                White   Female   
15057      Prof-specialty         Husband                White     Male   
15058        Adm-clerical       Own-child   Asian-Pac-Islander     Male   
15059     Exec-managerial         Husband                White     Male   

       capitalgain  capitalloss  hoursperweek          native  Salary  
0                0            0            40   United-States       0  
1                0            0            50   United-States       0  
2                0            0            40   United-States       1  
3             7688            0            40   United-States       1  
4                0            0            30   United-States       0  
...            ...          ...           ...             ...     ...  
15055            0            0            40   United-States       0  
15056            0            0            36   United-States       0  
15057            0            0            50   United-States       0  
15058         5455            0            40   United-States       0  
15059            0            0            60   United-States       1  

[15060 rows x 14 columns]

In [26]:
data_test_x=pd.get_dummies(data_test.iloc[:,:13])
data_test_x

age  educationno  capitalgain  capitalloss  hoursperweek  \
0       25            7            0            0            40   
1       38            9            0            0            50   
2       28           12            0            0            40   
3       44           10         7688            0            40   
4       34            6            0            0            30   
...    ...          ...          ...          ...           ...   
15055   33           13            0            0            40   
15056   39           13            0            0            36   
15057   38           13            0            0            50   
15058   44           13         5455            0            40   
15059   35           13            0            0            60   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
0                           0                     0                   1   
1                           0                     0                   1   
2                           0                     1                   0   
3                           0                     0                   1   
4                           0                     0                   1   
...                       ...                   ...                 ...   
15055                       0                     0                   1   
15056                       0                     0                   1   
15057                       0                     0                   1   
15058                       0                     0                   1   
15059                       0                     0                   0   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  ...  \
0                            0                            0  ...   
1                            0                            0  ...   
2                            0                            0  ...   
3                            0                            0  ...   
4                            0                            0  ...   
...                        ...                          ...  ...   
15055                        0                            0  ...   
15056                        0                            0  ...   
15057                        0                            0  ...   
15058                        0                            0  ...   
15059                        1                            0  ...   

       native_ Portugal  native_ Puerto-Rico  native_ Scotland  native_ South  \
0                     0                    0                 0              0   
1                     0                    0                 0              0   
2                     0                    0                 0              0   
3                     0                    0                 0              0   
4                     0                    0                 0              0   
...                 ...                  ...               ...            ...   
15055                 0                    0                 0              0   
15056                 0                    0                 0              0   
15057                 0                    0                 0              0   
15058                 0                    0                 0              0   
15059                 0                    0                 0              0   

       native_ Taiwan  native_ Thailand  native_ Trinadad&Tobago  \
0                   0                 0                        0   
1                   0                 0                        0   
2                   0                 0                        0   
3                   0                 0                        0   
4                   0                 0                        0   
...               ...               ...                      ...   
15055               0              

In [27]:
X_test=data_test_x
Y_test=data_test['Salary']

In [28]:
Y_test_pred=model_final.predict(X_test)

In [29]:
confusion_matrix(Y_test, Y_test_pred)

array([[11319,    41],
       [ 3017,   683]], dtype=int64)

In [30]:
acc = accuracy_score(Y_test, Y_test_pred) * 100
print("Accuracy =", acc)

Accuracy = 79.6945551128818


For the given test dataset we got accuracy of 79.7% using SVM model we created